### PR TITLE
refactor(cvi): use condition builder as condition api in cvi controller

### PR DESCRIPTION
**Pull request by eofff**
## Description
Use condition builder as condition api in cvi controller.

## Checklist
- [ ] The code is covered by unit tests.
- [x] e2e tests passed.
- [ ] Documentation updated according to the changes.
- [x] Changes were tested in the Kubernetes cluster manually.

### CODE REVIEW

File: images/virtualization-artifact/pkg/controller/cvi/internal/source/http.go [hunk 1]
```diff
@@ -59,8 +60,9 @@ func NewHTTPDataSource(
 func (ds HTTPDataSource) Sync(ctx context.Context, cvi *virtv2.ClusterVirtualImage) (reconcile.Result, error) {
 	log, ctx := logger.GetDataSourceContext(ctx, "http")
 
-	condition, _ := service.GetCondition(cvicondition.ReadyType, cvi.Status.Conditions)
-	defer func() { service.SetCondition(condition, &cvi.Status.Conditions) }()
+	condition, _ := conditions.GetCondition(cvicondition.ReadyType, cvi.Status.Conditions)
+	cb := conditions.NewConditionBuilder(cvicondition.ReadyType).FromCondition(condition).Generation(cvi.Generation)
+	defer func() { conditions.SetCondition(cb, &cvi.Status.Conditions) }()
```
Comment (Isteb4k) [inline images/virtualization-artifact/pkg/controller/cvi/internal/source/http.go:65]: Will result in an error if reason is not set

**Review by Isteb4k — CHANGES_REQUESTED** [review]

File: images/virtualization-artifact/pkg/controller/cvi/internal/source/http.go (hunk 1 above)
Comment (eofff) [reply images/virtualization-artifact/pkg/controller/cvi/internal/source/http.go:65]: It's logic problem, not builder

File: images/virtualization-artifact/pkg/controller/cvi/internal/datasource_ready.go [hunk 2]
```diff
@@ -42,16 +42,9 @@ func NewDatasourceReadyHandler(sources *source.Sources) *DatasourceReadyHandler
...
 func (h DatasourceReadyHandler) Handle(ctx context.Context, cvi *virtv2.ClusterVirtualImage) (reconcile.Result, error) {
-	condition, ok := service.GetCondition(cvicondition.DatasourceReadyType, cvi.Status.Conditions)
-	if !ok {
-		condition = metav1.Condition{
-			Type:   cvicondition.DatasourceReadyType,
-			Status: metav1.ConditionUnknown,
-			Reason: conditions.ReasonUnknown.String(),
-		}
-	}
+	cb := conditions.NewConditionBuilder(cvicondition.DatasourceReadyType).Generation(cvi.Generation)
```
Comment (yaroslavborbat) [inline images/virtualization-artifact/pkg/controller/cvi/internal/datasource_ready.go:45]: Add Status Unknown

File: images/virtualization-artifact/pkg/controller/cvi/internal/datasource_ready.go [hunk 3]
```diff
@@ -60,41 +53,47 @@ func (h DatasourceReadyHandler) Handle(ctx context.Context, cvi *virtv2.ClusterV
 	s, ok := h.sources.Get(cvi.Spec.DataSource.Type)
 	if !ok {
 		err := fmt.Errorf("data source validator not found for type: %s", cvi.Spec.DataSource.Type)
-		condition.Message = err.Error()
+		cb.Message(err.Error()) // TODO: check this line
```
Comment (yaroslavborbat) [inline images/virtualization-artifact/pkg/controller/cvi/internal/datasource_ready.go:56]: Why err, but not plain text

File: images/virtualization-artifact/pkg/controller/cvi/internal/datasource_ready.go [hunk 4]
```diff
@@ -60,41 +53,47 @@ func (h DatasourceReadyHandler) Handle(ctx context.Context, cvi *virtv2.ClusterV
...
 	err := s.Validate(ctx, cvi)
 	switch {
 	case err == nil:
-		condition.Status = metav1.ConditionTrue
-		condition.Reason = cvicondition.DatasourceReady
-		condition.Message = ""
+		cb.
+			Status(metav1.ConditionTrue).
+			Reason(cvicondition.DatasourceReady).
+			Message("")
```
Comment (yaroslavborbat) [inline images/virtualization-artifact/pkg/controller/cvi/internal/datasource_ready.go:66]: is there a need?

File: images/virtualization-artifact/pkg/controller/cvi/internal/source/http.go [hunk 5]
```diff
@@ -59,8 +60,9 @@ func NewHTTPDataSource(
 func (ds HTTPDataSource) Sync(ctx context.Context, cvi *virtv2.ClusterVirtualImage) (reconcile.Result, error) {
 	log, ctx := logger.GetDataSourceContext(ctx, "http")
 
-	condition, _ := service.GetCondition(cvicondition.ReadyType, cvi.Status.Conditions)
-	defer func() { service.SetCondition(condition, &cvi.Status.Conditions) }()
+	condition, _ := conditions.GetCondition(cvicondition.ReadyType, cvi.Status.Conditions)
+	cb := conditions.NewConditionBuilder(cvicondition.ReadyType).Generation(cvi.Generation)
```
Comment (yaroslavborbat) [inline images/virtualization-artifact/pkg/controller/cvi/internal/source/http.go:64]: Add Status Unknown

File: images/virtualization-artifact/pkg/controller/cvi/internal/source/object_ref.go [hunk 6]
```diff
@@ -72,8 +73,9 @@ func NewObjectRefDataSource(
 func (ds ObjectRefDataSource) Sync(ctx context.Context, cvi *virtv2.ClusterVirtualImage) (reconcile.Result, error) {
 	log, ctx := logger.GetDataSourceContext(ctx, "objectref")
 
-	condition, _ := service.GetCondition(cvicondition.ReadyType, cvi.Status.Conditions)
-	defer func() { service.SetCondition(condition, &cvi.Status.Conditions) }()
+	condition, _ := conditions.GetCondition(cvicondition.ReadyType, cvi.Status.Conditions)
+	cb := conditions.NewConditionBuilder(cvicondition.ReadyType).Generation(cvi.Generation)
```
Comment (yaroslavborbat) [inline images/virtualization-artifact/pkg/controller/cvi/internal/source/object_ref.go:77]: Unknown?

Comment (yaroslavborbat) [reply images/virtualization-artifact/pkg/controller/cvi/internal/source/object_ref.go:77]: and Reason

File: images/virtualization-artifact/pkg/controller/vi/internal/datasource_ready.go [hunk 7]
```diff
@@ -94,7 +93,7 @@ func (h DatasourceReadyHandler) Handle(ctx context.Context, vi *virtv2.VirtualIm
 		return reconcile.Result{}, nil
 	case errors.As(err, &source.VirtualDiskAttachedToRunningVMError{}):
 		condition.Status = metav1.ConditionFalse
-		condition.Reason = cvicondition.VirtualDiskNotReady
+		condition.Reason = vicondition.VirtualDiskNotReady
 		condition.Message = service.CapitalizeFirstLetter(err.Error() + ".")
```
Comment (yaroslavborbat) [inline images/virtualization-artifact/pkg/controller/vi/internal/datasource_ready.go:97]: cb?

**Review by yaroslavborbat — CHANGES_REQUESTED** [review]

File: images/virtualization-artifact/pkg/controller/cvi/internal/life_cycle.go [hunk 8]
```diff
@@ -44,10 +44,10 @@ func NewLifeCycleHandler(sources *source.Sources, client client.Client) *LifeCyc
 }
 
 func (h LifeCycleHandler) Handle(ctx context.Context, cvi *virtv2.ClusterVirtualImage) (reconcile.Result, error) {
-	readyCondition, ok := service.GetCondition(cvicondition.ReadyType, cvi.Status.Conditions)
+	readyCondition, ok := service.GetCondition(cvicondition.ReadyType.String(), cvi.Status.Conditions)
```
Comment (Isteb4k) [inline images/virtualization-artifact/pkg/controller/cvi/internal/life_cycle.go:47]: Use conditions.GetCondition
And at the line 55 conditions.SetCondition

**Review by Isteb4k — CHANGES_REQUESTED** [review]

**Review by yaroslavborbat — DISMISSED** [review]

**Review by Isteb4k — APPROVED** [review]

**Pull request merged**